### DISCUSSION TITLE
Relax the restriction of pattern matching on generic constraints

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#7020] [Language Ideas]
This is a state. Most of them only decide what to do when switching
```csharp
	public interface IState<TSelf> where TSelf : class, IState<TSelf>
	{
		void Enter(TSelf state);

		void Exit(TSelf state);
	}
```
Then they have some extensions
```csharp
	public interface IActivityState<TSelf> : IState<TSelf> where TSelf : class, IState<TSelf>
	{
		public void Activity();
	}
```
```csharp
	public interface IStateMachine<in TState> where TState : class, ISwitchState<TState>
	{
		public void SwitchOn<T>() where T : TState;
	}

	public interface ISwitchState<TSelf> : IState<TSelf> where TSelf : class, ISwitchState<TSelf>
	{
		void Init(IStateMachine<TSelf> machine);
	}
```
This is my default state machine. He will continue to do some things himself. At the same time, he will try to execute the activity in the state
```csharp
	public sealed class StateMachine2<TState> : IStateMachine<TState> where TState : class, ISwitchState<TState>
	{
		private TState CurrentState;

		public void Activity()
		{
			if (CurrentState is IActivityState<TState> activity)
			{
				activity.Activity();
			}
//something
		}
		public void SwitchOn<T>() where T : TState => throw new NotImplementedException();
	}
```
The problem now is that the self-constraint of the extension interface only constrains the base type. However, if you change it to its own state, the state machine cannot perform pattern matching because it cannot guarantee that the constraint is satisfied.
```
public interface IActivityState<TSelf> : IState<TSelf> where TSelf : class, IActivityState<TSelf>

if (CurrentState is IActivityState<TState> activity)//CS0311
```
I hope that as long as I think it is possible to match the pattern successfully, let me pass the compilation. For example, it is obviously possible here.

**Comment by svick** [comment]
I think this proposal would also do what you're looking for: https://github.com/dotnet/csharplang/discussions/6308.
